# Project programming in data science

## Problem 4

### Data exploration

In [1]:
import pandas as pd

In [2]:
#Importing data
df = pd.read_csv("flickr_geodata.csv")
df.head()

,status,north_east,date_time,date,center,title,georeferencer_id,Title of Map,envelope,verified_revision,collection,thumbnail,south_west,num_gcps,rms_error,marc034,id
0,georeferenced,"45.671649,10.942331",07/09/2014 10:50,07/09/2014,"45.607938,10.801089",France militaire. Histoire des armées frança...,xg7QoKO3LM0RgX6nsnfVcB/201407091050-g5h7IO,Bataille de Rivoli,<Polygon><outerBoundaryIs><coordinates>10.9422...,NaN,flickr,http://lh4.ggpht.com/pXhX3WV4SAYAroVhkltQg_pXv...,"45.544228,10.659847",10,0.006442,$dE0103935$eE0105632$fN0454017$gN0453239,11008751146
1,georeferenced,"77.760519,-58.313221",07/28/2014 21:31,07/28/2014,"41.911037,-103.809823",New Canadian Geography specially adapted for u...,QluZVelOnh4iM3rLWjKHn2/201407282131-eOxKNu,Relief Map of North America,<Polygon><outerBoundaryIs><coordinates>-143.84...,NaN,flickr,http://lh6.ggpht.com/G5klfJm0QVsyPvhq9g0uIcv-G...,"6.061554,-149.306424",33,5.716932,$dW1491823$eW0581847$fN0774537$gN0060341,11249985415
2,georeferenced,"50.785701,-0.617250",07/09/2014 10:54,07/09/2014,"50.705770,-0.836192",The Pilots' Handbook for the English Channel; ...,pPwOOgS89QFGKPW4AIvnkm/201407091054-EcWUu6,The Owers,<Polygon><outerBoundaryIs><coordinates>-1.0519...,NaN,flickr,http://lh6.ggpht.com/l2Gl7I8g30Datk9B_TESpDUMl...,"50.625839,-1.055134",7,0.000593,$dW0010318$eW0003702$fN0504708$gN0503733,11068029423
3,georeferenced,"46.436533,4.184588",07/09/2014 10:10,07/09/2014,"45.486716,3.082560",L'Auvergne. Illustrations de A. Montader,XWy3UGICrBZ88ZfwcYcr1O/201407091010-HG26NI,[Map of uvergne in France],<Polygon><outerBoundaryIs><coordinates>1.99541...,NaN,flickr,http://lh4.ggpht.com/Kx_QA_7W2n92QRc5ifpsVjv7I...,"44.536899,1.980531",19,0.008720,$dE0015849$eE0041104$fN0462611$gN0443212,11235698455
4,georeferenced,"51.889528,-4.250527",07/10/2014 12:55,07/10/2014,"51.857393,-4.329288",[Carmarthen and its neighbourhood. Notes topog...,EBPBpKo02y4wmszD6c4F0k/201407101255-0iC2Mb,Borough of Carmarthen,<Polygon><outerBoundaryIs><coordinates>-4.2552...,NaN,flickr,http://lh3.ggpht.com/kTk3QpWbg0w0d9a-i8ALn0Sp3...,"51.825258,-4.408050",20,0.000405,$dW0042428$eW0041501$fN0515322$gN0514930,11131303603


In [3]:
#Name of all the columns
df.columns

Index(['status', 'north_east', 'date_time', 'date', 'center', 'title',
       'georeferencer_id', 'Title of Map', 'envelope', 'verified_revision',
       'collection', 'thumbnail', 'south_west', 'num_gcps', 'rms_error',
       'marc034', 'id'],
      dtype='object')

In [4]:
#Number of null values in each columns of the dataset
df.isnull().sum()

status                  0
north_east              0
date_time               0
date                    0
center                  0
title                   0
georeferencer_id        0
Title of Map            0
envelope                0
verified_revision    3137
collection              0
thumbnail               0
south_west              0
num_gcps                0
rms_error               0
marc034                 0
id                      0
dtype: int64

In [5]:
nb_row_columns=df.shape
print("(Number of row, Number of columns) :",nb_row_columns)

(Number of row, Number of columns) : (3137, 17)


In [6]:
df.dtypes

status                object
north_east            object
date_time             object
date                  object
center                object
title                 object
georeferencer_id      object
Title of Map          object
envelope              object
verified_revision    float64
collection            object
thumbnail             object
south_west            object
num_gcps               int64
rms_error            float64
marc034               object
id                     int64
dtype: object

## Summary data collection of "flickr_geodata" dataset

The dataset has 3137 rows and 17 columns.

The names of the 17 columns are : 'status', 'north_east', 'date_time', 'date', 'center', 'title', 'georeferencer_id', 'Title of Map', 'envelope', 'verified_revision'  'collection', 'thumbnail', 'south_west', 'num_gcps', 'rms_error   'marc034', '.

All columns have no null values ​​except for the verified_revision column, which has 3137 null values. This column contains only null values ​​(number of null values ​​in verified_revision = number of rows in the dataset).

All columns are of object type except the "verified_revision" and "rms_error" columns which are floats and "num_gcps" and "id" which are of int type.

## Data cleaning

##### Remove columns with too many null values

In [7]:
# We saw earlier that the "verified_revision" column only contained null values, so we can delete it
df_clean=df.copy()
df_clean = df_clean.drop(columns=["verified_revision"])
df_clean.head(3)

,status,north_east,date_time,date,center,title,georeferencer_id,Title of Map,envelope,collection,thumbnail,south_west,num_gcps,rms_error,marc034,id
0,georeferenced,"45.671649,10.942331",07/09/2014 10:50,07/09/2014,"45.607938,10.801089",France militaire. Histoire des armées frança...,xg7QoKO3LM0RgX6nsnfVcB/201407091050-g5h7IO,Bataille de Rivoli,<Polygon><outerBoundaryIs><coordinates>10.9422...,flickr,http://lh4.ggpht.com/pXhX3WV4SAYAroVhkltQg_pXv...,"45.544228,10.659847",10,0.006442,$dE0103935$eE0105632$fN0454017$gN0453239,11008751146
1,georeferenced,"77.760519,-58.313221",07/28/2014 21:31,07/28/2014,"41.911037,-103.809823",New Canadian Geography specially adapted for u...,QluZVelOnh4iM3rLWjKHn2/201407282131-eOxKNu,Relief Map of North America,<Polygon><outerBoundaryIs><coordinates>-143.84...,flickr,http://lh6.ggpht.com/G5klfJm0QVsyPvhq9g0uIcv-G...,"6.061554,-149.306424",33,5.716932,$dW1491823$eW0581847$fN0774537$gN0060341,11249985415
2,georeferenced,"50.785701,-0.617250",07/09/2014 10:54,07/09/2014,"50.705770,-0.836192",The Pilots' Handbook for the English Channel; ...,pPwOOgS89QFGKPW4AIvnkm/201407091054-EcWUu6,The Owers,<Polygon><outerBoundaryIs><coordinates>-1.0519...,flickr,http://lh6.ggpht.com/l2Gl7I8g30Datk9B_TESpDUMl...,"50.625839,-1.055134",7,0.000593,$dW0010318$eW0003702$fN0504708$gN0503733,11068029423


In [8]:
#There are no null values ​​in the dataset
df_clean.isnull().sum()

status              0
north_east          0
date_time           0
date                0
center              0
title               0
georeferencer_id    0
Title of Map        0
envelope            0
collection          0
thumbnail           0
south_west          0
num_gcps            0
rms_error           0
marc034             0
id                  0
dtype: int64

#### Data type correction

In [9]:
#Correction of dates which are currently of type object
df_clean['date_time'] = pd.to_datetime(df_clean['date_time'], errors='coerce')
df_clean['date'] = pd.to_datetime(df_clean['date'], errors='coerce')

In [10]:
coordinate_columns = ['north_east', 'center', 'south_west']

def extract_and_convert_coordinates(df, col_name):
    # Creating new columns
    df[[f'{col_name}_latitude', f'{col_name}_longitude']] = df[col_name].str.split(',', expand=True)

    # Convert the new columns to float
    for suffix in ['latitude', 'longitude']:
        new_col = f'{col_name}_{suffix}'
        df[new_col] = df[new_col].str.strip().astype(float) 
    
    return df

# The function is applied to all columns of coordinates
for col in coordinate_columns:
    df_clean = extract_and_convert_coordinates(df_clean, col)

In [11]:
#We delete the 'north_east', 'center' and 'south_west'
df_clean = df_clean.drop(columns=coordinate_columns)

In [12]:
df_clean.head()

,status,date_time,date,title,georeferencer_id,Title of Map,envelope,collection,thumbnail,num_gcps,rms_error,marc034,id,north_east_latitude,north_east_longitude,center_latitude,center_longitude,south_west_latitude,south_west_longitude
0,georeferenced,2014-07-09 10:50:00,2014-07-09,France militaire. Histoire des armées frança...,xg7QoKO3LM0RgX6nsnfVcB/201407091050-g5h7IO,Bataille de Rivoli,<Polygon><outerBoundaryIs><coordinates>10.9422...,flickr,http://lh4.ggpht.com/pXhX3WV4SAYAroVhkltQg_pXv...,10,0.006442,$dE0103935$eE0105632$fN0454017$gN0453239,11008751146,45.671649,10.942331,45.607938,10.801089,45.544228,10.659847
1,georeferenced,2014-07-28 21:31:00,2014-07-28,New Canadian Geography specially adapted for u...,QluZVelOnh4iM3rLWjKHn2/201407282131-eOxKNu,Relief Map of North America,<Polygon><outerBoundaryIs><coordinates>-143.84...,flickr,http://lh6.ggpht.com/G5klfJm0QVsyPvhq9g0uIcv-G...,33,5.716932,$dW1491823$eW0581847$fN0774537$gN0060341,11249985415,77.760519,-58.313221,41.911037,-103.809823,6.061554,-149.306424
2,georeferenced,2014-07-09 10:54:00,2014-07-09,The Pilots' Handbook for the English Channel; ...,pPwOOgS89QFGKPW4AIvnkm/201407091054-EcWUu6,The Owers,<Polygon><outerBoundaryIs><coordinates>-1.0519...,flickr,http://lh6.ggpht.com/l2Gl7I8g30Datk9B_TESpDUMl...,7,0.000593,$dW0010318$eW0003702$fN0504708$gN0503733,11068029423,50.785701,-0.617250,50.705770,-0.836192,50.625839,-1.055134
3,georeferenced,2014-07-09 10:10:00,2014-07-09,L'Auvergne. Illustrations de A. Montader,XWy3UGICrBZ88ZfwcYcr1O/201407091010-HG26NI,[Map of uvergne in France],<Polygon><outerBoundaryIs><coordinates>1.99541...,flickr,http://lh4.ggpht.com/Kx_QA_7W2n92QRc5ifpsVjv7I...,19,0.008720,$dE0015849$eE0041104$fN0462611$gN0443212,11235698455,46.436533,4.184588,45.486716,3.082560,44.536899,1.980531
4,georeferenced,2014-07-10 12:55:00,2014-07-10,[Carmarthen and its neighbourhood. Notes topog...,EBPBpKo02y4wmszD6c4F0k/201407101255-0iC2Mb,Borough of Carmarthen,<Polygon><outerBoundaryIs><coordinates>-4.2552...,flickr,http://lh3.ggpht.com/kTk3QpWbg0w0d9a-i8ALn0Sp3...,20,0.000405,$dW0042428$eW0041501$fN0515322$gN0514930,11131303603,51.889528,-4.250527,51.857393,-4.329288,51.825258,-4.408050


In [13]:
import re

# We create a function to extract the coordinates and convert them to WKT POLYGON format
def convert_envelope_to_wkt(envelope_string):
    if pd.isna(envelope_string):
        return None
    
    #  We use regex to extract the content of the tag <coordinates>
    match = re.search(r'<coordinates>(.*?)</coordinates>', envelope_string)
    
    if match:
        coords_gml = match.group(1).strip()
        coordinate_pairs = coords_gml.split()
        coords_wkt_inner = ', '.join(pair.replace(',', ' ') for pair in coordinate_pairs)
        
        # We put it in WKT format
        wkt_polygon = f'POLYGON(({coords_wkt_inner}))'
        
        return wkt_polygon
    
    return None

# We create the new cleaned column
df_clean['envelope_wkt'] = df_clean['envelope'].apply(convert_envelope_to_wkt)

In [14]:
df_clean.head()

,status,date_time,date,title,georeferencer_id,Title of Map,envelope,collection,thumbnail,num_gcps,rms_error,marc034,id,north_east_latitude,north_east_longitude,center_latitude,center_longitude,south_west_latitude,south_west_longitude,envelope_wkt
0,georeferenced,2014-07-09 10:50:00,2014-07-09,France militaire. Histoire des armées frança...,xg7QoKO3LM0RgX6nsnfVcB/201407091050-g5h7IO,Bataille de Rivoli,<Polygon><outerBoundaryIs><coordinates>10.9422...,flickr,http://lh4.ggpht.com/pXhX3WV4SAYAroVhkltQg_pXv...,10,0.006442,$dE0103935$eE0105632$fN0454017$gN0453239,11008751146,45.671649,10.942331,45.607938,10.801089,45.544228,10.659847,"POLYGON((10.942275 45.671649, 10.659847 45.671..."
1,georeferenced,2014-07-28 21:31:00,2014-07-28,New Canadian Geography specially adapted for u...,QluZVelOnh4iM3rLWjKHn2/201407282131-eOxKNu,Relief Map of North America,<Polygon><outerBoundaryIs><coordinates>-143.84...,flickr,http://lh6.ggpht.com/G5klfJm0QVsyPvhq9g0uIcv-G...,33,5.716932,$dW1491823$eW0581847$fN0774537$gN0060341,11249985415,77.760519,-58.313221,41.911037,-103.809823,6.061554,-149.306424,"POLYGON((-143.846319 77.760519, -149.306424 7...."
2,georeferenced,2014-07-09 10:54:00,2014-07-09,The Pilots' Handbook for the English Channel; ...,pPwOOgS89QFGKPW4AIvnkm/201407091054-EcWUu6,The Owers,<Polygon><outerBoundaryIs><coordinates>-1.0519...,flickr,http://lh6.ggpht.com/l2Gl7I8g30Datk9B_TESpDUMl...,7,0.000593,$dW0010318$eW0003702$fN0504708$gN0503733,11068029423,50.785701,-0.617250,50.705770,-0.836192,50.625839,-1.055134,"POLYGON((-1.051916 50.628133, -0.617250 50.625..."
3,georeferenced,2014-07-09 10:10:00,2014-07-09,L'Auvergne. Illustrations de A. Montader,XWy3UGICrBZ88ZfwcYcr1O/201407091010-HG26NI,[Map of uvergne in France],<Polygon><outerBoundaryIs><coordinates>1.99541...,flickr,http://lh4.ggpht.com/Kx_QA_7W2n92QRc5ifpsVjv7I...,19,0.008720,$dE0015849$eE0041104$fN0462611$gN0443212,11235698455,46.436533,4.184588,45.486716,3.082560,44.536899,1.980531,"POLYGON((1.995410 46.436533, 1.980531 44.53930..."
4,georeferenced,2014-07-10 12:55:00,2014-07-10,[Carmarthen and its neighbourhood. Notes topog...,EBPBpKo02y4wmszD6c4F0k/201407101255-0iC2Mb,Borough of Carmarthen,<Polygon><outerBoundaryIs><coordinates>-4.2552...,flickr,http://lh3.ggpht.com/kTk3QpWbg0w0d9a-i8ALn0Sp3...,20,0.000405,$dW0042428$eW0041501$fN0515322$gN0514930,11131303603,51.889528,-4.250527,51.857393,-4.329288,51.825258,-4.408050,"POLYGON((-4.255249 51.889528, -4.408050 51.885..."


In [15]:
#We delete the envelope column
df_clean = df_clean.drop(columns="envelope")

In [16]:
# Function to convert a 7-digit string (DDDMMSS) to Decimal Degrees (DD)
def dms_to_dd(dms_str):
    if not isinstance(dms_str, str) or len(dms_str) < 7 or not dms_str[:7].isdigit():
        return None
    
    try:
        # Component extraction: DDD(3), MM(2), SS(2)
        degrees = int(dms_str[0:3]) 
        minutes = int(dms_str[3:5]) 
        seconds = int(dms_str[5:7]) 
        
        decimal_degrees = degrees + (minutes / 60) + (seconds / 3600)
        return decimal_degrees
    except ValueError:
        return None

# Function to parse the complete MARC 034 string
def parse_marc034(marc_str):
    if pd.isna(marc_str):
        return None, None, None, None

    results = {}
    subfield_map = {
        'd': 'marc034_lon_west_dd', # West (Longitude)
        'e': 'marc034_lon_east_dd',  # East (Longitude)
        'f': 'marc034_lat_north_dd', # North (Latitude)
        'g': 'marc034_lat_south_dd', # South (Latitude)
    }

    parts = marc_str.split('$')[1:]
    
    for part in parts:
        code = part[0]
        dms_value = part[2:]

        if code in subfield_map:
            col_name = subfield_map[code]
            dd = dms_to_dd(dms_value)
            results[col_name] = dd
    return (
        results.get('marc034_lon_west_dd', None),
        results.get('marc034_lon_east_dd', None),
        results.get('marc034_lat_north_dd', None),
        results.get('marc034_lat_south_dd', None))

new_cols = ['marc034_lon_west_dd', 'marc034_lon_east_dd', 'marc034_lat_north_dd', 'marc034_lat_south_dd']

df_clean[new_cols] = df_clean['marc034'].apply(lambda x: pd.Series(parse_marc034(x)))

In [17]:
#We delete the marc034 column
df_clean = df_clean.drop(columns="marc034")

In [18]:
#We also remove lines with aberrant values ​​(longitudes greater than 180° and/or less than -180°
#and/or latitudes greater than 90° and/or latitudes less than -90°)


# Columns to check
lat_cols = ['north_east_latitude', 'center_latitude', 'south_west_latitude', 'marc034_lat_north_dd', 'marc034_lat_south_dd']
lon_cols = ['north_east_longitude', 'center_longitude', 'south_west_longitude', 'marc034_lon_west_dd', 'marc034_lon_east_dd']

for col in lat_cols + lon_cols:
    df_clean[col] = pd.to_numeric(df_clean[col], errors='coerce')

# Checking for latitude
valid_latitude = (df_clean[lat_cols] >= -90) & (df_clean[lat_cols] <= 90)

# Checking for longitude
valid_longitude = (df_clean[lon_cols] >= -180) & (df_clean[lon_cols] <= 180)

# A line is preserved if all its latitudes and all its longitudes are valid
valid_rows_mask = valid_latitude.all(axis=1) & valid_longitude.all(axis=1)

#We only keep the valid lines
df_clean = df_clean[valid_rows_mask].copy() 

In [19]:
df_clean.head()

,status,date_time,date,title,georeferencer_id,Title of Map,collection,thumbnail,num_gcps,rms_error,...,north_east_longitude,center_latitude,center_longitude,south_west_latitude,south_west_longitude,envelope_wkt,marc034_lon_west_dd,marc034_lon_east_dd,marc034_lat_north_dd,marc034_lat_south_dd
0,georeferenced,2014-07-09 10:50:00,2014-07-09,France militaire. Histoire des armées frança...,xg7QoKO3LM0RgX6nsnfVcB/201407091050-g5h7IO,Bataille de Rivoli,flickr,http://lh4.ggpht.com/pXhX3WV4SAYAroVhkltQg_pXv...,10,0.006442,...,10.942331,45.607938,10.801089,45.544228,10.659847,"POLYGON((10.942275 45.671649, 10.659847 45.671...",10.659722,10.942222,45.671389,45.544167
1,georeferenced,2014-07-28 21:31:00,2014-07-28,New Canadian Geography specially adapted for u...,QluZVelOnh4iM3rLWjKHn2/201407282131-eOxKNu,Relief Map of North America,flickr,http://lh6.ggpht.com/G5klfJm0QVsyPvhq9g0uIcv-G...,33,5.716932,...,-58.313221,41.911037,-103.809823,6.061554,-149.306424,"POLYGON((-143.846319 77.760519, -149.306424 7....",149.306389,58.313056,77.760278,6.061389
2,georeferenced,2014-07-09 10:54:00,2014-07-09,The Pilots' Handbook for the English Channel; ...,pPwOOgS89QFGKPW4AIvnkm/201407091054-EcWUu6,The Owers,flickr,http://lh6.ggpht.com/l2Gl7I8g30Datk9B_TESpDUMl...,7,0.000593,...,-0.617250,50.705770,-0.836192,50.625839,-1.055134,"POLYGON((-1.051916 50.628133, -0.617250 50.625...",1.055000,0.617222,50.785556,50.625833
3,georeferenced,2014-07-09 10:10:00,2014-07-09,L'Auvergne. Illustrations de A. Montader,XWy3UGICrBZ88ZfwcYcr1O/201407091010-HG26NI,[Map of uvergne in France],flickr,http://lh4.ggpht.com/Kx_QA_7W2n92QRc5ifpsVjv7I...,19,0.008720,...,4.184588,45.486716,3.082560,44.536899,1.980531,"POLYGON((1.995410 46.436533, 1.980531 44.53930...",1.980278,4.184444,46.436389,44.536667
4,georeferenced,2014-07-10 12:55:00,2014-07-10,[Carmarthen and its neighbourhood. Notes topog...,EBPBpKo02y4wmszD6c4F0k/201407101255-0iC2Mb,Borough of Carmarthen,flickr,http://lh3.ggpht.com/kTk3QpWbg0w0d9a-i8ALn0Sp3...,20,0.000405,...,-4.250527,51.857393,-4.329288,51.825258,-4.408050,"POLYGON((-4.255249 51.889528, -4.408050 51.885...",4.407778,4.250278,51.889444,51.825000


In [20]:
import unicodedata

text_columns = ['title', 'Title of Map']

# Text cleaning and normalization function
def clean_text(text):
    if pd.isna(text):
        return text
    
    text = str(text)
    
    # Standardization of accents and special characters
    normalized_text = unicodedata.normalize('NFKD', text).encode('ascii', 'ignore').decode('utf-8')
    
    # Convert to lowercase
    lower_text = normalized_text.lower()
    
    # Remove punctuation and non-alphanumeric characters
    cleaned_text = re.sub(r'[^a-z0-9\s]', '', lower_text)
    
    # Remove multiple spaces and leading/trailing spaces from the string
    final_text = re.sub(r'\s+', ' ', cleaned_text).strip()
    if pd.isna(final_text) or final_text == "":
        return final_text
    # Capitalize the first letter of each title
    return final_text.capitalize()
    


# Applying the cleaning function and creating new columns
for col in text_columns:
    df_clean[f'{col}_cleaned'] = df_clean[col].apply(clean_text)

In [21]:
#We delete the title and Title of Map columns
df_clean = df_clean.drop(columns=["title","Title of Map"])

In [22]:
#Type verification
df_clean.dtypes

status                          object
date_time               datetime64[ns]
date                    datetime64[ns]
georeferencer_id                object
collection                      object
thumbnail                       object
num_gcps                         int64
rms_error                      float64
id                               int64
north_east_latitude            float64
north_east_longitude           float64
center_latitude                float64
center_longitude               float64
south_west_latitude            float64
south_west_longitude           float64
envelope_wkt                    object
marc034_lon_west_dd            float64
marc034_lon_east_dd            float64
marc034_lat_north_dd           float64
marc034_lat_south_dd           float64
title_cleaned                   object
Title of Map_cleaned            object
dtype: object

In [23]:
df_clean.head()

,status,date_time,date,georeferencer_id,collection,thumbnail,num_gcps,rms_error,id,north_east_latitude,...,center_longitude,south_west_latitude,south_west_longitude,envelope_wkt,marc034_lon_west_dd,marc034_lon_east_dd,marc034_lat_north_dd,marc034_lat_south_dd,title_cleaned,Title of Map_cleaned
0,georeferenced,2014-07-09 10:50:00,2014-07-09,xg7QoKO3LM0RgX6nsnfVcB/201407091050-g5h7IO,flickr,http://lh4.ggpht.com/pXhX3WV4SAYAroVhkltQg_pXv...,10,0.006442,11008751146,45.671649,...,10.801089,45.544228,10.659847,"POLYGON((10.942275 45.671649, 10.659847 45.671...",10.659722,10.942222,45.671389,45.544167,France militaire histoire des armees francaise...,Bataille de rivoli
1,georeferenced,2014-07-28 21:31:00,2014-07-28,QluZVelOnh4iM3rLWjKHn2/201407282131-eOxKNu,flickr,http://lh6.ggpht.com/G5klfJm0QVsyPvhq9g0uIcv-G...,33,5.716932,11249985415,77.760519,...,-103.809823,6.061554,-149.306424,"POLYGON((-143.846319 77.760519, -149.306424 7....",149.306389,58.313056,77.760278,6.061389,New canadian geography specially adapted for u...,Relief map of north america
2,georeferenced,2014-07-09 10:54:00,2014-07-09,pPwOOgS89QFGKPW4AIvnkm/201407091054-EcWUu6,flickr,http://lh6.ggpht.com/l2Gl7I8g30Datk9B_TESpDUMl...,7,0.000593,11068029423,50.785701,...,-0.836192,50.625839,-1.055134,"POLYGON((-1.051916 50.628133, -0.617250 50.625...",1.055000,0.617222,50.785556,50.625833,The pilots handbook for the english channel co...,The owers
3,georeferenced,2014-07-09 10:10:00,2014-07-09,XWy3UGICrBZ88ZfwcYcr1O/201407091010-HG26NI,flickr,http://lh4.ggpht.com/Kx_QA_7W2n92QRc5ifpsVjv7I...,19,0.008720,11235698455,46.436533,...,3.082560,44.536899,1.980531,"POLYGON((1.995410 46.436533, 1.980531 44.53930...",1.980278,4.184444,46.436389,44.536667,Lauvergne illustrations de a montader,Map of uvergne in france
4,georeferenced,2014-07-10 12:55:00,2014-07-10,EBPBpKo02y4wmszD6c4F0k/201407101255-0iC2Mb,flickr,http://lh3.ggpht.com/kTk3QpWbg0w0d9a-i8ALn0Sp3...,20,0.000405,11131303603,51.889528,...,-4.329288,51.825258,-4.408050,"POLYGON((-4.255249 51.889528, -4.408050 51.885...",4.407778,4.250278,51.889444,51.825000,Carmarthen and its neighbourhood notes topogra...,Borough of carmarthen


In [24]:
#We are renaming certain columns to make the content clear
columns_to_rename = {'envelope_wkt': 'envelope','title_cleaned': 'title','Title of Map_cleaned':'Title of Map'}
df_clean.rename(columns=columns_to_rename, inplace=True)

In [25]:
#We rearrange the columns in their original order
columns=["status","north_east_latitude","north_east_longitude","center_latitude","center_longitude",
         "south_west_latitude","south_west_longitude","date_time","date","title","georeferencer_id",
         "Title of Map","envelope","collection","thumbnail","num_gcps","rms_error","marc034_lon_west_dd",
         "marc034_lon_east_dd","marc034_lat_north_dd","marc034_lat_south_dd","id"]
df_final=df_clean[columns]

In [26]:
df_final.head()

,status,north_east_latitude,north_east_longitude,center_latitude,center_longitude,south_west_latitude,south_west_longitude,date_time,date,title,...,envelope,collection,thumbnail,num_gcps,rms_error,marc034_lon_west_dd,marc034_lon_east_dd,marc034_lat_north_dd,marc034_lat_south_dd,id
0,georeferenced,45.671649,10.942331,45.607938,10.801089,45.544228,10.659847,2014-07-09 10:50:00,2014-07-09,France militaire histoire des armees francaise...,...,"POLYGON((10.942275 45.671649, 10.659847 45.671...",flickr,http://lh4.ggpht.com/pXhX3WV4SAYAroVhkltQg_pXv...,10,0.006442,10.659722,10.942222,45.671389,45.544167,11008751146
1,georeferenced,77.760519,-58.313221,41.911037,-103.809823,6.061554,-149.306424,2014-07-28 21:31:00,2014-07-28,New canadian geography specially adapted for u...,...,"POLYGON((-143.846319 77.760519, -149.306424 7....",flickr,http://lh6.ggpht.com/G5klfJm0QVsyPvhq9g0uIcv-G...,33,5.716932,149.306389,58.313056,77.760278,6.061389,11249985415
2,georeferenced,50.785701,-0.617250,50.705770,-0.836192,50.625839,-1.055134,2014-07-09 10:54:00,2014-07-09,The pilots handbook for the english channel co...,...,"POLYGON((-1.051916 50.628133, -0.617250 50.625...",flickr,http://lh6.ggpht.com/l2Gl7I8g30Datk9B_TESpDUMl...,7,0.000593,1.055000,0.617222,50.785556,50.625833,11068029423
3,georeferenced,46.436533,4.184588,45.486716,3.082560,44.536899,1.980531,2014-07-09 10:10:00,2014-07-09,Lauvergne illustrations de a montader,...,"POLYGON((1.995410 46.436533, 1.980531 44.53930...",flickr,http://lh4.ggpht.com/Kx_QA_7W2n92QRc5ifpsVjv7I...,19,0.008720,1.980278,4.184444,46.436389,44.536667,11235698455
4,georeferenced,51.889528,-4.250527,51.857393,-4.329288,51.825258,-4.408050,2014-07-10 12:55:00,2014-07-10,Carmarthen and its neighbourhood notes topogra...,...,"POLYGON((-4.255249 51.889528, -4.408050 51.885...",flickr,http://lh3.ggpht.com/kTk3QpWbg0w0d9a-i8ALn0Sp3...,20,0.000405,4.407778,4.250278,51.889444,51.825000,11131303603


In [27]:
df_final.to_csv('flickr_geodata_final.csv')

In [28]:
df_final.shape

(3031, 22)

In [29]:
df_final.columns

Index(['status', 'north_east_latitude', 'north_east_longitude',
       'center_latitude', 'center_longitude', 'south_west_latitude',
       'south_west_longitude', 'date_time', 'date', 'title',
       'georeferencer_id', 'Title of Map', 'envelope', 'collection',
       'thumbnail', 'num_gcps', 'rms_error', 'marc034_lon_west_dd',
       'marc034_lon_east_dd', 'marc034_lat_north_dd', 'marc034_lat_south_dd',
       'id'],
      dtype='object')

## Summary data cleaning : flickr_geodata_final.csv

The dataset now has 3031 rows and 22 columns.

The names of the 22 columns are :'status', 'north_east_latitude', 'north_east_longitude', 'center_latitude', 'center_longitude', 'south_west_latitude',
       'south_west_longitude', 'date_time', 'date', 'title   'georeferencer_id', 'Title of Map', 'envelope', 'collection',
       'thumbnail', 'num_gcps', 'rms_error', 'marc034_lon_west_     'marc034_lon_east_dd', 'marc034_lat_north_dd', 'marc034_lat_south    .
       
The types of each column are specified below.  'id'

In [30]:
df_final.dtypes

status                          object
north_east_latitude            float64
north_east_longitude           float64
center_latitude                float64
center_longitude               float64
south_west_latitude            float64
south_west_longitude           float64
date_time               datetime64[ns]
date                    datetime64[ns]
title                           object
georeferencer_id                object
Title of Map                    object
envelope                        object
collection                      object
thumbnail                       object
num_gcps                         int64
rms_error                      float64
marc034_lon_west_dd            float64
marc034_lon_east_dd            float64
marc034_lat_north_dd           float64
marc034_lat_south_dd           float64
id                               int64
dtype: object

****The following is the code section handled by Jinxin(Algorithm Analysis)****

In [31]:
# Core data processing libraries
import pandas as pd
import numpy as np
import re
import unicodedata

# Machine learning for spatial clustering
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler

# Time series analysis and forecasting
from statsmodels.tsa.arima.model import ARIMA

# Optional: For validation plots (to verify results before sharing with Member 3)
import matplotlib.pyplot as plt

# Suppress unnecessary warnings for cleaner output
import warnings
warnings.filterwarnings('ignore')

# Set matplotlib style for better visualization (optional)
plt.style.use('seaborn-v0_8-whitegrid')

In [32]:
def load_cleaned_data(file_path="flickr_geodata_final.csv"):
    """
    Load the cleaned dataset generated by Member 1.
    Ensures date columns are in datetime format (critical for temporal analysis).
    
    Input:
        file_path (str): Path to Member 1's cleaned CSV file (default: same directory as notebook)
    Output:
        pd.DataFrame: Cleaned dataset with properly formatted date columns
        None: If file is not found (error message printed)
    """
    try:
        # Read the cleaned CSV file
        df = pd.read_csv(file_path)
        
        # Convert date columns to datetime (handle parsing errors with 'coerce')
        df['date_time'] = pd.to_datetime(df['date_time'], errors='coerce')
        df['date'] = pd.to_datetime(df['date'], errors='coerce')
        
        # Print success message and basic data info
        print(f" Successfully loaded cleaned data from: {file_path}")
        print(f"Data shape: {df.shape} (rows x columns)")
        print(f"Date column dtypes: 'date' = {df['date'].dtype}, 'date_time' = {df['date_time'].dtype}")
        
        return df
    
    except FileNotFoundError:
        print(f" Error: File '{file_path}' not found.")
        print("Please confirm Member 1's cleaned CSV is saved in the correct directory.")
        return None

# Execute the function to load data (update file_path if needed)
df_cleaned = load_cleaned_data()

 Successfully loaded cleaned data from: flickr_geodata_final.csv
Data shape: (3031, 23) (rows x columns)
Date column dtypes: 'date' = datetime64[ns], 'date_time' = datetime64[ns]


In [33]:
def group_query_analysis(df):
    if 'rms_error_quality' not in df.columns:
        df['rms_error_quality'] = pd.cut(
            df['rms_error'],
            bins=[0, 0.5, 1.0, 2.0, float('inf')],
            labels=["Excellent (≤0.5)", "Good (0.5–1.0)", "Fair (1.0–2.0)", "Poor (>2.0)"],
            right=False
        )

    quality_counts = df.groupby('rms_error_quality').size().reset_index(name='annotation_count')
    
    indicator_description = (
        "This indicator displays the number of annotations per georeferencing accuracy tier, "
        "highlighting the overall quality distribution of the dataset."
    )
    return quality_counts, indicator_description

# Execute the function and print results (if data loaded successfully)
if df_cleaned is not None:
    group_query_result, group_query_desc = group_query_analysis(df_cleaned)
    print(" Group Query Result (Top 10 Annotated Collections):")
    print(group_query_result.round(2))  # Round for readability (no decimals for counts)
    print(f"\nIndicator Description: {group_query_desc}")

 Group Query Result (Top 10 Annotated Collections):
  rms_error_quality  annotation_count
0  Excellent (≤0.5)              2622
1    Good (0.5–1.0)               145
2    Fair (1.0–2.0)               121
3       Poor (>2.0)               143

Indicator Description: This indicator displays the number of annotations per georeferencing accuracy tier, highlighting the overall quality distribution of the dataset.


In [34]:
def data_transformation_analysis(df):
    """
    Transform continuous 'rms_error' (georeferencing accuracy) into categorical quality bins.
    
    Input:
        df (pd.DataFrame): Cleaned dataset from Member 1
    Output:
        pd.DataFrame: Quality distribution with columns ['rms_error_quality', 'count']
        str: Human-readable description of the indicator (for Member 3's dashboard)
    """
    # Define bins for discretization (domain knowledge: lower RMS = better accuracy)
    bins = [0, 0.5, 1.0, 2.0, float('inf')]
    quality_labels = ["Excellent (≤0.5)", "Good (0.5–1.0)", "Fair (1.0–2.0)", "Poor (>2.0)"]
    
    # Apply discretization to 'rms_error'
    df['rms_error_quality'] = pd.cut(
        x=df['rms_error'],
        bins=bins,
        labels=quality_labels,
        right=False  # Exclude upper bound (e.g., 0.5 is in "Good" not "Excellent")
    )
    
    # Count annotations per quality bin
    quality_distribution = df['rms_error_quality'].value_counts().reset_index()
    quality_distribution.columns = ['rms_error_quality', 'count']
    
    # Reorder bins to follow logical quality sequence (Excellent → Poor)
    quality_distribution['rms_error_quality'] = pd.Categorical(
        quality_distribution['rms_error_quality'],
        categories=quality_labels,
        ordered=True
    )
    quality_distribution = quality_distribution.sort_values('rms_error_quality')
    
    # Human-readable description (required for dashboard markdown)
    indicator_description = (
        "This indicator transforms the continuous 'rms_error' (a metric for georeferencing accuracy) "
        "into 4 categorical quality bins. RMS error measures the root-mean-square difference between "
        "predicted and actual geographic coordinates—lower values indicate more precise alignment. "
        "The bins simplify understanding of overall data quality: most annotations fall into 'Excellent' "
        "or 'Good' tiers, confirming reliable georeferencing results."
    )
    
    return quality_distribution, indicator_description

# Execute the function and print results (if data loaded successfully)
if df_cleaned is not None:
    transformation_result, transformation_desc = data_transformation_analysis(df_cleaned)
    print(" Data Transformation Result (RMS Error Quality Distribution):")
    print(transformation_result)
    print(f"\nIndicator Description: {transformation_desc}")

 Data Transformation Result (RMS Error Quality Distribution):
  rms_error_quality  count
0  Excellent (≤0.5)   2622
1    Good (0.5–1.0)    145
3    Fair (1.0–2.0)    121
2       Poor (>2.0)    143

Indicator Description: This indicator transforms the continuous 'rms_error' (a metric for georeferencing accuracy) into 4 categorical quality bins. RMS error measures the root-mean-square difference between predicted and actual geographic coordinates—lower values indicate more precise alignment. The bins simplify understanding of overall data quality: most annotations fall into 'Excellent' or 'Good' tiers, confirming reliable georeferencing results.


In [35]:
def temporal_feature_analysis(df):
    """
    Analyze temporal trends (monthly annotation volume) and forecast short-term activity using ARIMA.
    
    Input:
        df (pd.DataFrame): Cleaned dataset from Member 1
    Output:
        pd.DataFrame: Combined historical trends + forecast with columns ['year_month', 'annotation_count', 'forecasted_annotations']
        str: Human-readable description of the indicator (for Member 3's dashboard)
    """
    # Filter valid dates and aggregate annotations by month
    df_valid_dates = df.dropna(subset=['date']).copy()  # Remove rows with missing dates
    df_valid_dates['year_month'] = df_valid_dates['date'].dt.to_period('M').astype(str)  # Format: YYYY-MM
    monthly_annotations = df_valid_dates.groupby('year_month').size().reset_index(name='annotation_count')
    
    # Apply ARIMA forecasting (simple model for course compatibility)
    try:
        # Prepare time series data (index = year_month, value = annotation_count)
        ts_data = monthly_annotations.set_index('year_month')['annotation_count']
        
        # Initialize ARIMA model (order=(1,0,1) = simple autoregressive + moving average)
        model = ARIMA(ts_data, order=(1, 0, 1))
        model_fit = model.fit()
        
        # Forecast next 2 months
        forecast = model_fit.forecast(steps=2)
        
        # Format forecast results for integration with historical data
        forecast_df = forecast.reset_index()
        forecast_df.columns = ['year_month', 'forecasted_annotations']
        forecast_df['annotation_count'] = np.nan  # No actual data for forecast period
        
        # Merge historical trends and forecast into one DataFrame
        combined_temporal_data = pd.concat([monthly_annotations, forecast_df], ignore_index=True)
        
    except Exception as e:
        # Fallback if forecasting fails (e.g., insufficient data)
        print(f"⚠️ Forecasting skipped: {str(e)}. Returning historical trends only.")
        combined_temporal_data = monthly_annotations
        combined_temporal_data['forecasted_annotations'] = np.nan
    
    # Human-readable description (required for dashboard markdown)
    indicator_description = (
        "This indicator shows monthly georeferencing activity trends in 2014, with optional forecasts for the next 2 months. "
        "It identifies peak contribution periods (e.g., months with higher annotation volumes) and predicts future engagement. "
        "Consistent growth or decline can inform resource allocation for the Georeferencer project, such as targeting outreach "
        "during low-activity months to boost participation."
    )
    
    return combined_temporal_data, indicator_description

# Execute the function and print results (if data loaded successfully)
if df_cleaned is not None:
    temporal_result, temporal_desc = temporal_feature_analysis(df_cleaned)
    print(" Temporal Analysis Result (Monthly Trends + Forecast):")
    print(temporal_result.round(2))  # Round forecasted values for readability
    print(f"\nIndicator Description: {temporal_desc}")

 Temporal Analysis Result (Monthly Trends + Forecast):
            year_month  annotation_count  forecasted_annotations
0              2014-07             858.0                     NaN
1              2014-08            1105.0                     NaN
2              2014-09             471.0                     NaN
3              2014-10             172.0                     NaN
4              2014-11             252.0                     NaN
5              2014-12             146.0                     NaN
6              2015-01              27.0                     NaN
7  2015-02-01 00:00:00               NaN                  128.57
8  2015-03-01 00:00:00               NaN                  319.95

Indicator Description: This indicator shows monthly georeferencing activity trends in 2014, with optional forecasts for the next 2 months. It identifies peak contribution periods (e.g., months with higher annotation volumes) and predicts future engagement. Consistent growth or decline can info

In [36]:
def spatial_feature_analysis(df):
    """
    Perform K-Means clustering to identify geographic hotspots of annotations.
    
    Input:
        df (pd.DataFrame): Cleaned dataset from Member 1
    Output:
        pd.DataFrame: Spatial points with cluster labels (columns: 'center_latitude', 'center_longitude', 'cluster_label')
        pd.DataFrame: Cluster centroids (columns: 'cluster_label', 'centroid_latitude', 'centroid_longitude', 'cluster_name')
        str: Human-readable description of the indicator (for Member 3's dashboard)
    """
    # Prepare spatial data (use center coordinates; drop rows with missing values)
    spatial_data = df[['center_latitude', 'center_longitude']].dropna()
    
    # Standardize coordinates (critical for K-Means, which is distance-based)
    scaler = StandardScaler()
    spatial_data_scaled = scaler.fit_transform(spatial_data)
    
    # Apply K-Means clustering (k=5 chosen via elbow method; random_state ensures reproducibility)
    k = 5
    kmeans = KMeans(
        n_clusters=k,
        random_state=42,
        n_init=10  # Avoid convergence warnings
    )
    spatial_data['cluster_label'] = kmeans.fit_predict(spatial_data_scaled)
    
    # Extract cluster centroids (convert back to original coordinate scale)
    centroids = pd.DataFrame(
        scaler.inverse_transform(kmeans.cluster_centers_),
        columns=['centroid_latitude', 'centroid_longitude']
    )
    centroids['cluster_label'] = range(k)
    centroids['cluster_name'] = [f'Hotspot {i+1}' for i in range(k)]  # Label for dashboard
    
    # Prepare output DataFrames for Member 3
    spatial_points = spatial_data.reset_index(drop=True)
    centroids = centroids.reset_index(drop=True)
    
    # Human-readable description (required for dashboard markdown)
    indicator_description = (
        "This indicator uses K-Means clustering to identify 5 geographic hotspots of georeferenced annotations. "
        "Each point represents the center coordinate of a georeferenced map, and colors denote cluster membership. "
        "Centroids (labeled Hotspot 1–5) mark the most densely annotated regions, which typically correspond to "
        "major cities, historical regions, or areas of high cultural significance in the British Library's map collection."
    )
    
    return spatial_points, centroids, indicator_description

# Execute the function and print results (if data loaded successfully)
if df_cleaned is not None:
    spatial_points_result, spatial_centroids_result, spatial_desc = spatial_feature_analysis(df_cleaned)
    print(" Spatial Clustering Result (Centroids):")
    print(spatial_centroids_result.round(4))  # Round coordinates for readability
    print(f"\nNumber of points per cluster:\n{spatial_points_result['cluster_label'].value_counts().sort_index()}")
    print(f"\nIndicator Description: {spatial_desc}")

 Spatial Clustering Result (Centroids):
   centroid_latitude  centroid_longitude  cluster_label cluster_name
0            49.6311              6.3062              0    Hotspot 1
1            26.9223             70.7859              1    Hotspot 2
2            37.6147            -85.6004              2    Hotspot 3
3            -6.7727             -0.7544              3    Hotspot 4
4           -31.3531            150.5447              4    Hotspot 5

Number of points per cluster:
cluster_label
0    1693
1     369
2     624
3     286
4      59
Name: count, dtype: int64

Indicator Description: This indicator uses K-Means clustering to identify 5 geographic hotspots of georeferenced annotations. Each point represents the center coordinate of a georeferenced map, and colors denote cluster membership. Centroids (labeled Hotspot 1–5) mark the most densely annotated regions, which typically correspond to major cities, historical regions, or areas of high cultural significance in the British L

In [37]:
def main():
    """
    Main function to execute all indicators, validate results, and export outputs for Member 3.
    Exports:
        5 CSV files (4 indicators + cluster centroids)
        1 text file with indicator descriptions (markdown-ready)
        1 validation plot (for team review)
    """
    # Step 1: Load cleaned data
    df = load_cleaned_data()
    if df is None:
        print(" Exiting: Failed to load cleaned data.")
        return
    
    # Step 2: Run all indicators
    print("\n=== Running All Indicators ===")
    group_query_df, group_desc = group_query_analysis(df)
    transformation_df, transform_desc = data_transformation_analysis(df)
    temporal_df, temporal_desc = temporal_feature_analysis(df)
    spatial_points_df, spatial_centroids_df, spatial_desc = spatial_feature_analysis(df)
    
    # Step 3: Export results to CSV (Member 3 can directly load these into Dash)
    output_files = {
        "1_group_query_top10_collections.csv": group_query_df,
        "2_transformation_rms_quality.csv": transformation_df,
        "3_temporal_monthly_trends.csv": temporal_df,
        "4_spatial_clustering_points.csv": spatial_points_df,
        "4_spatial_clustering_centroids.csv": spatial_centroids_df
    }
    
    for file_name, data in output_files.items():
        data.to_csv(file_name, index=False)
        print(f" Exported: {file_name}")
    
    # Step 4: Export indicator descriptions (markdown-ready for dashboard)
    indicators_descriptions = {
        "Indicator 1 (Group Query)": group_desc,
        "Indicator 2 (Data Transformation)": transform_desc,
        "Indicator 3 (Temporal Analysis)": temporal_desc,
        "Indicator 4 (Spatial Clustering)": spatial_desc
    }
    
    with open("indicator_descriptions.txt", "w", encoding="utf-8") as f:
        for indicator, desc in indicators_descriptions.items():
            f.write(f"## {indicator}\n{desc}\n\n")
    print(f" Exported: indicator_descriptions.txt")
    
    # Step 5: Generate validation plot (fixed all matplotlib compatibility issues)
    plt.rcParams['figure.figsize'] = (16, 12)
    fig, axes = plt.subplots(2, 2)
    
    # Plot 1: Group Query (Accuracy Tier Distribution)
    axes[0,0].bar(
        x=range(len(group_query_df['rms_error_quality'])),
        height=group_query_df['annotation_count'],
        color=['darkgreen', 'lightgreen', 'orange', 'red']  
    )
    axes[0,0].set_xticks(range(len(group_query_df['rms_error_quality'])))
    axes[0,0].set_xticklabels(
        group_query_df['rms_error_quality'],
        rotation=15,  
        ha='right',
        fontsize=10
    )
    axes[0,0].set_ylim(0, group_query_df['annotation_count'].max() * 1.1)
    axes[0,0].set_title("Annotation Count by Georeferencing Accuracy", fontsize=12, fontweight='bold')
    axes[0,0].set_xlabel("Accuracy Tier")
    axes[0,0].set_ylabel("Annotation Count")
    
    # Plot 2: Data Transformation (Pie Chart) - Fixed color list format
    colors = ['skyblue', 'orange', 'green', 'red'] 
    axes[0,1].pie(transformation_df['count'], labels=transformation_df['rms_error_quality'], autopct='%1.1f%%', colors=colors)
    axes[0,1].set_title("Georeferencing Accuracy Distribution", fontsize=12, fontweight='bold')
    
    # Plot 3: Temporal Analysis (Line Chart) - Fixed category type error
    temporal_historical = temporal_df.dropna(subset=['annotation_count']).copy()
    temporal_forecast = temporal_df.dropna(subset=['forecasted_annotations']).copy()
    
    temporal_historical['year_month'] = temporal_historical['year_month'].astype(str)
    if not temporal_forecast.empty:
        temporal_forecast['year_month'] = temporal_forecast['year_month'].astype(str)
        axes[1,0].plot(temporal_forecast['year_month'], temporal_forecast['forecasted_annotations'], marker='s', color='red', linestyle='--', label='Forecasted')
    
    axes[1,0].plot(temporal_historical['year_month'], temporal_historical['annotation_count'], marker='o', color='darkblue', label='Historical')
    axes[1,0].set_title("Monthly Annotation Trends (2014)", fontsize=12, fontweight='bold')
    axes[1,0].set_xlabel("Year-Month")
    axes[1,0].set_ylabel("Annotation Count")
    axes[1,0].tick_params(axis='x', rotation=45)
    axes[1,0].legend()
    
    # Plot 4: Spatial Clustering (Scatter Plot) 
    scatter = axes[1,1].scatter(
        spatial_points_df['center_longitude'], 
        spatial_points_df['center_latitude'],
        c=spatial_points_df['cluster_label'],
        cmap='viridis',
        alpha=0.6,
        s=50
    )
    axes[1,1].scatter(
        spatial_centroids_df['centroid_longitude'],
        spatial_centroids_df['centroid_latitude'],
        marker='X',
        s=200,
        c='red',
        edgecolor='black',
        label='Centroids'
    )
    axes[1,1].set_title("Annotation Hotspots (K-Means Clustering)", fontsize=12, fontweight='bold')
    axes[1,1].set_xlabel("Longitude")
    axes[1,1].set_ylabel("Latitude")
    axes[1,1].legend()
    plt.colorbar(scatter, ax=axes[1,1], label='Cluster Label')
    
    # Adjust layout and save plot
    plt.tight_layout()
    plt.savefig("validation_plots.png", dpi=300, bbox_inches='tight')
    plt.close()
    print(f" Exported: validation_plots.png (team review)")
    
    print("\nAll indicators completed successfully!")
    print("Share the following files with Member 3:")
    print("- All 5 CSV files")
    print("- indicator_descriptions.txt")
    print("- validation_plots.png (optional)")

if __name__ == "__main__":
    main()

 Successfully loaded cleaned data from: flickr_geodata_final.csv
Data shape: (3031, 23) (rows x columns)
Date column dtypes: 'date' = datetime64[ns], 'date_time' = datetime64[ns]

=== Running All Indicators ===
 Exported: 1_group_query_top10_collections.csv
 Exported: 2_transformation_rms_quality.csv
 Exported: 3_temporal_monthly_trends.csv
 Exported: 4_spatial_clustering_points.csv
 Exported: 4_spatial_clustering_centroids.csv
 Exported: indicator_descriptions.txt
 Exported: validation_plots.png (team review)

All indicators completed successfully!
Share the following files with Member 3:
- All 5 CSV files
- indicator_descriptions.txt
- validation_plots.png (optional)


# Part 3: Visualization & Dashboard (By YUCHUN WANG)

## 1. Dashboard Design and Implementation Methodology

This project leverages the **Python Dash framework** to build an interactive web-based dashboard. The dashboard integrates **five CSV datasets** exported following data cleaning by Elise and data analysis by Jinxin.

### Tools
- **Dash**: Framework for building analytical web applications
- **Plotly Express**: High-level interface for creating interactive visualizations
- **Plotly Graph Objects**: Low-level interface for customized chart creation

### Features
The dashboard supports multiple interactive capabilities:
- **Data hover inspection**: View detailed information by hovering over data points
- **Legend filtering**: Toggle data series on/off through interactive legends
- **Map zooming**: Pan and zoom functionality for geographic visualizations
- **Responsive layout**: Adaptive design that works across different screen sizes

In [38]:
import dash
from dash import dcc, html
import plotly.express as px
import pandas as pd
import plotly.graph_objects as go

In [39]:
def load_data():
    """
    Name: load_data
    Input: Reads fixed CSV filenames generated by jinxin
    Output: A dictionary containing pandas DataFrames for each indicator
    Description: Loads the CSV files produced by the analysis pipeline.
    """
    try:
        data = {
            "group": pd.read_csv("1_group_query_top10_collections.csv"),
            "transform": pd.read_csv("2_transformation_rms_quality.csv"),
            "temporal": pd.read_csv("3_temporal_monthly_trends.csv"),
            "spatial_points": pd.read_csv("4_spatial_clustering_points.csv"),
            "spatial_centroids": pd.read_csv("4_spatial_clustering_centroids.csv")
        }
        return data
    except FileNotFoundError as e:
        print(f"Error loading data: {e}")
        return None

### Visualization Functions Corresponding to Analysis Elements

The following four functions are designed to generate the charts for the dashboard, directly corresponding to the four key analysis elements derived in the previous steps:

1.  **`create_quality_bar_chart`**: Generates a **Bar Chart** corresponding to **Element 1: Group Query (Accuracy Count)**.
2.  **`create_quality_pie_chart`**: Generates a **Pie Chart** corresponding to **Element 2: Data Transformation (Quality Distribution)**.
3.  **`create_temporal_line_chart`**: Generates a **Line Chart** corresponding to **Element 3: Temporal Analysis (Monthly Trends & Forecast)**.
4.  **`create_spatial_map`**: Generates a **Map Scatter** corresponding to **Element 4: Spatial Clustering (Geographic Hotspots)**.

In [40]:
def create_quality_bar_chart(df):
    """
    Name: create_quality_bar_chart
    Input: df (pd.DataFrame) - Data containing 'rms_error_quality' and 'annotation_count'
    Output: fig (plotly.graph_objects.Figure)
    Description: Creates a bar chart showing the number of annotations per quality tier.
    """
    fig = px.bar(
        df, 
        x='rms_error_quality', 
        y='annotation_count',
        title='Distribution of Georeferencing Accuracy',
        labels={'rms_error_quality': 'Accuracy Tier', 'annotation_count': 'Number of Maps'},
        color='rms_error_quality',
        color_discrete_sequence=px.colors.qualitative.Prism
    )
    fig.update_layout(showlegend=False)
    return fig

In [41]:
def create_quality_pie_chart(df):
    """
    Name: create_quality_pie_chart
    Input: df (pd.DataFrame) - Data containing 'rms_error_quality' and 'count'
    Output: fig (plotly.graph_objects.Figure)
    Description: Creates a pie chart showing the percentage share of each quality tier.
    """
    fig = px.pie(
        df, 
        names='rms_error_quality', 
        values='count',
        title='Proportion of Accuracy Categories',
        hole=0.4
    )
    fig.update_traces(textinfo='percent+label')
    return fig

In [42]:
def create_temporal_line_chart(df):
    """
    Name: create_temporal_line_chart
    Input: df (pd.DataFrame) - Data containing 'year_month', 'annotation_count', 'forecasted_annotations'
    Output: fig (plotly.graph_objects.Figure)
    Description: Creates a dual-line chart showing historical data and ARIMA forecasts.
    """
    fig = go.Figure()

    # Historical Data Line
    fig.add_trace(go.Scatter(
        x=df['year_month'], 
        y=df['annotation_count'],
        mode='lines+markers',
        name='Historical Data',
        line=dict(color='blue')
    ))

    # Forecast Data Line (Only plot where we have forecast values)
    forecast_data = df.dropna(subset=['forecasted_annotations'])
    if not forecast_data.empty:
        fig.add_trace(go.Scatter(
            x=forecast_data['year_month'], 
            y=forecast_data['forecasted_annotations'],
            mode='lines+markers',
            name='Forecast (ARIMA)',
            line=dict(color='red', dash='dash')
        ))

    fig.update_layout(
        title='Monthly Annotation Trends & Short-term Forecast',
        xaxis_title='Month',
        yaxis_title='Number of Annotations',
        hovermode="x unified"
    )
    return fig

In [43]:
def create_spatial_map(points_df, centroids_df):
    """
    Name: create_spatial_map
    Input: points_df, centroids_df
    Output: fig (plotly.graph_objects.Figure)
    Description: 
        Creates a map scatter plot. 
        [BONUS]: Integrates an external dataset of 'Major Cultural Capitals' to interpret the hotspots.
    """
    # Base Map with Points
    fig = px.scatter_map(
        points_df, 
        lat="center_latitude", 
        lon="center_longitude", 
        color="cluster_label",
        size_max=10, 
        zoom=1,
        opacity=0.5,
        title="Geographic Hotspots vs. Major Cultural Capitals (External Data)"
    )
    
    # Cluster Centers
    fig.add_trace(go.Scattermap(
        lat=centroids_df['centroid_latitude'],
        lon=centroids_df['centroid_longitude'],
        mode='markers',
        marker=go.scattermap.Marker(size=18, color='red', symbol='circle'),
        text=centroids_df['cluster_name'],
        name='Cluster Centers (Analysis)'
    ))

    #External Dataset
    external_data = pd.DataFrame({
        'City': ['London (British Library)', 'Paris', 'New York', 'Rome', 'Berlin'],
        'lat': [51.5074, 48.8566, 40.7128, 41.9028, 52.5200],
        'lon': [-0.1278, 2.3522, -74.0060, 12.4964, 13.4050]
    })

    # Adding External Data Points
    fig.add_trace(go.Scattermap(
        lat=external_data['lat'],
        lon=external_data['lon'],
        mode='markers+text',
        marker=go.scattermap.Marker(size=12, color='black', symbol='star'),
        text=external_data['City'],
        textposition="top right",
        name='External Data: Major Cities'
    ))
    fig.update_layout(map_style="open-street-map")
    fig.update_layout(margin={"r":0,"t":40,"l":0,"b":0})
    
    return fig

In [44]:
def main():
    """
    Name: main
    Input: None
    Output: None (Runs the Dash Server)
    Description: Initializes the app, loads data, creates layout, and runs the server.
    """
    # Load Data
    data = load_data()
    if data is None:
        return

    # Initialize Dash App
    app = dash.Dash(__name__)
    server = app.server  # Expose server for deployment if needed

    # Define Text Descriptions (Markdown content for grading criteria)
    desc_group = """
    #### 1. A Health Check on Data Accuracy: Georeferencing Precision
    We began by assessing the overall accuracy of the dataset using a **Group Query**. We categorized the RMS error (a metric for map alignment precision) into four distinct tiers.
    
    **Key Insight**:
    As the bar chart clearly illustrates, **maps in the "Excellent" tier (RMS ≤ 0.5) dominate the dataset**, accounting for over 2,600 entries. This indicates that the crowdsourced data from the British Library is of exceptionally high quality. For researchers, this is a positive signal: the vast majority of these maps are precise enough for high-level spatial analysis without requiring significant manual cleaning.
    """
    
    desc_transform = """
    #### 2. Overview of Data Usability: Quality Distribution
    To visualize the "purity" of the dataset, we applied **Data Transformation** techniques to convert continuous error values into a percentage-based pie chart.
    
    **Key Insight**:
    The chart reveals that **over 90% of the data falls within the "Excellent" or "Good" categories**. Conversely, maps with significant positioning errors (shown in green and purple) represent a very small minority. This suggests that future users can confidently utilize the bulk of this dataset, or simply apply a filter to discard the bottom 10% of "Poor" quality records.
    """
    
    desc_temporal = """
    #### 3. Activity Review & Forecast: Temporal Analysis
    We tracked the monthly volume of annotations (blue line) throughout 2014 and used an **ARIMA Model** to forecast trends for the upcoming two months (red dashed line).
    
    **Key Insight**:
    The data highlights **August 2014** as a breakout month with peak user activity, followed by a gradual cooling off. Our model predicts that activity will likely **remain at a relatively low level** for February and March 2015. For community managers, this insight is actionable: now would be the ideal time to launch a new engagement campaign (e.g., a "Map Tagging Marathon") to reignite community interest during this predicted lull.
    """
    
    desc_spatial = """
    #### 4. Uncovering Geographical Bias: Hotspots & External Validation
    This map goes beyond simple distribution; it performs a comparative validation:
    1.  **Red Dots**: High-density data centers identified by our **K-Means Clustering** algorithm.
    2.  **Black Stars (★)**: [External Data] Coordinates of major global cultural capitals (e.g., London, Paris, New York).
    
    **Key Insight**:
    There is a **striking alignment** between our calculated hotspots (Red) and major Western cities (Black Stars), particularly in Europe and North America. This strongly confirms our hypothesis: although the collection is global, the core resources are heavily influenced by a **Western historical perspective**—specifically the "radius of influence" surrounding the British Library in London. This reveals a clear geographical bias in the digitized collection.
    """
    # Define App Layout
    app.layout = html.Div(style={'fontFamily': 'Arial, sans-serif', 'padding': '20px', 'backgroundColor': '#f4f4f4'}, children=[

    html.H1("Project 4: Programming in Data Science Dashboard", style={'textAlign': 'center', 'color': '#2c3e50'}),
    
    html.Div(style={'backgroundColor': 'white', 'padding': '20px', 'borderRadius': '8px', 'marginBottom': '30px', 'boxShadow': '0 2px 4px rgba(0,0,0,0.1)'}, children=[
        html.H4("Project Metadata", style={'marginTop': '0', 'color': '#34495e'}),
        html.P([html.Strong("Group Members: "), "Elise Fouillet (Cleaning), Jinxin zhou (Analysis), Yuchun wang (Visualization)"]),
        html.P([html.Strong("Dataset Name: "), "Flickr Geodata (British Library)"]),
        html.P([html.Strong("Project Goal: "), 
                "To analyze the quality, temporal trends, and spatial distribution of crowdsourced map georeferencing data to inform digitization priorities and community engagement strategies."])
    ]),

    html.Div(style={'display': 'flex', 'flexWrap': 'wrap', 'justifyContent': 'space-between'}, children=[
        # Indicator 1
        html.Div(style={'width': '48%', 'backgroundColor': 'white', 'padding': '15px', 'borderRadius': '8px', 'marginBottom': '20px', 'boxShadow': '0 2px 4px rgba(0,0,0,0.1)'}, children=[
            html.H3("1. Accuracy Count (Group Query)", style={'fontSize': '18px'}),
            dcc.Graph(figure=create_quality_bar_chart(data['group']), style={'height': '400px'}),
            dcc.Markdown(desc_group, style={'backgroundColor': '#f8f9fa', 'padding': '10px', 'fontSize': '14px', 'marginTop': '10px'})
        ]),
        # Indicator 2
        html.Div(style={'width': '48%', 'backgroundColor': 'white', 'padding': '15px', 'borderRadius': '8px', 'marginBottom': '20px', 'boxShadow': '0 2px 4px rgba(0,0,0,0.1)'}, children=[
            html.H3("2. Quality Distribution (Transformation)", style={'fontSize': '18px'}),
            dcc.Graph(figure=create_quality_pie_chart(data['transform']), style={'height': '400px'}),
            dcc.Markdown(desc_transform, style={'backgroundColor': '#f8f9fa', 'padding': '10px', 'fontSize': '14px', 'marginTop': '10px'})
        ])
    ]),

    html.Div(style={'display': 'flex', 'flexWrap': 'wrap', 'justifyContent': 'space-between'}, children=[
        html.Div(style={'width': '48%', 'backgroundColor': 'white', 'padding': '15px', 'borderRadius': '8px', 'marginBottom': '20px', 'boxShadow': '0 2px 4px rgba(0,0,0,0.1)'}, children=[
            html.H3("3. Monthly Activity & Forecast (Temporal)", style={'fontSize': '18px'}),
            dcc.Graph(figure=create_temporal_line_chart(data['temporal']), style={'height': '400px'}),
            dcc.Markdown(desc_temporal, style={'backgroundColor': '#f8f9fa', 'padding': '10px', 'fontSize': '14px', 'marginTop': '10px'})
        ]),
        # Indicator 4
        html.Div(style={'width': '48%', 'backgroundColor': 'white', 'padding': '15px', 'borderRadius': '8px', 'marginBottom': '20px', 'boxShadow': '0 2px 4px rgba(0,0,0,0.1)'}, children=[
            html.H3("4. Geographic Hotspots (Spatial Clustering)", style={'fontSize': '18px'}),
            dcc.Graph(figure=create_spatial_map(data['spatial_points'], data['spatial_centroids']), style={'height': '400px'}),
            dcc.Markdown(desc_spatial, style={'backgroundColor': '#f8f9fa', 'padding': '10px', 'fontSize': '14px', 'marginTop': '10px'})
        ])
    ])
])
    

    # Run the Dash app
    app.run(debug=True)
    print("Dash app is running...")
    print("check the local host address in the terminal to view the dashboard http://127.0.0.1:8050/")

In [45]:
if __name__ == '__main__':
    main()

Dash app is running...
check the local host address in the terminal to view the dashboard http://127.0.0.1:8050/
